File: Week_0/3_private_encaptulaiton.cpp
#include <bits/stdc++.h>
using namespace std;

class Bank {
protected:
    // Protected attribute
    string _branch;

private:
    // Private attribute
    double __balance;

public:
    // Public attribute (for demonstration)
    string holder_name;

    // Constructor
    Bank(string holderName, double initialDeposit)
    {
        this->holder_name = holderName;
        _branch = "Panthapath";  // protected attribute
        this->__balance = initialDeposit;  // Assign the private balance
    }

    // Method to deposit money
    void deposit(double depositAmount) {
        __balance += depositAmount;
    }

    // Method to withdraw money
    double withdraw(double withdrawAmount) {
        if (withdrawAmount <= __balance) {
            __balance -= withdrawAmount;
            return withdrawAmount;
        } else {
            cout << "Not enough balance: " << withdrawAmount << endl;
            return 0.0;
        }
    }

    // Method to get the balance (public accessor)
    double get_balance() const {
        return __balance;
    }
};


int main()
{
    // Creating a Bank object
    Bank sayed("Sayed", 10000);

    // Depositing money
    sayed.deposit(5000);

    // Accessing public attribute
    cout << "Account holder: " << sayed.holder_name << endl;

    // Accessing private balance through the public method
    cout << "Balance: " << sayed.get_balance() << endl;

    return 0;
}

File: Week_0/3_private_encaptulaiton.py
class Bank:
    def __init__(self, holder_name, initial_deposit) -> None:
        self.holder_name = holder_name # public attribute
        self._branch = 'panthapath' # protected attribute
        self.__balance = initial_deposit # private attribute
   
    # encapsulation: hiding details operation
    def deposit(self, deposit_amount):
        self.__balance += deposit_amount
   
    def withdraw(self, withdraw_amount):
        if withdraw_amount < self.__balance:
            self.__balance = self.__balance - withdraw_amount
            return withdraw_amount
        else:
            return f'Not enough balance: {withdraw_amount}'
   
    def get_balance(self):
        return self.__balance


sayed = Bank('Sayed', 10000)
sayed.deposit(5000)


print(sayed.holder_name)    # accessing public attributes


print(sayed.get_balance())  # accessing protected attributes
